modules/c++/re/source/RegexPCRE.cpp

```diff
@@ -70,244 +181,145 @@ re::Regex& re::Regex::operator=(const re::Regex& rhs)
+                          size_t& end)
 {
-    int numMatches(0);
-    int result(0);
-    int startOffset(0);
-
-    // Clear the output vector
-    mOvector.assign(mOvector.size(), 0);
-    numMatches = pcre_exec(mPCRE,                             // the compiled pattern
-                           NULL,                              // no extra data
-                           matchString.c_str() + startIndex,  // the subject string
-                           matchString.length() - startIndex, // the subject length
-                           startOffset,                       // starting offset
-                           flags,                             // options
-                           &mOvector[0],                      // output vector
-                           OVECTOR_COUNT);                    // output vector size
-
-    result = numMatches;
-
-    if (result == 0)
-    {
-        numMatches = OVECTOR_COUNT / 3;
-    }
-
-    if (result >= 0)
-    {
-        if (((mOvector[0] + startIndex) +
-             (mOvector[1] - mOvector[0])) >
-            (int)matchString.length() )
-        {
-            result = PCRE_ERROR_NOMATCH;
-        }
-    }
+    ScopedMatchData matchData(mPCRE);
+    const size_t numMatches = matchData.match(matchString, startIndex, flags);
 
-    if (result >= 0)
-    {
-        // output vector start offset = i*2
-        // output vector end   offset = i*2+1
-        // i = 0
-        int index = mOvector[0] + startIndex;
-        int subStringLength = mOvector[1] - mOvector[0];
-        int subStringCheck = index + subStringLength;
-        if (subStringCheck > (int)matchString.length())
-        {
-            throw RegexException(Ctxt(FmtX("Search: Match substring out of range (%d,%d) for string of length %d", index, subStringCheck, matchString.length())));
-        }
-        return matchString.substr(index, subStringLength);
-    }
-    else if (result == PCRE_ERROR_NOMATCH)
+    if (numMatches > 0)
     {
-        return std::string("");
+        begin = matchData.getOutputVector()[0];
+        end = matchData.getOutputVector()[1];
+        return matchData.getMatch(matchString, 0);
     }
     else
     {
-        throw RegexException
-            (Ctxt(FmtX("Error in searching for %s", matchString.c_str())));
+        begin = end = 0;
+        return "";
     }
 }
 
-void re::Regex::searchAll(const std::string& matchString,
-                          RegexMatch & v)
+void Regex::searchAll(const std::string& matchString, RegexMatch& v)
 {
-    std::string result = search(matchString, 0, 0);
+    size_t startIndex = 0;
+
+    size_t begin;
+    size_t end;
+    std::string result = search(matchString, startIndex, 0, begin, end);
 
-    int idx = 0;
-    while (result.size() != 0)
+    while (!result.empty())
     {
         v.push_back(result);
-        idx += (int)result.size() + mOvector[0];
-        result = search(matchString, idx, PCRE_NOTBOL);
+        startIndex = end;
```

Comment:
Is `end` the last matching letter, or one beyond?  If it's the last matching, shouldn't this be `startIndex = end + 1;`?  Won't make a difference unless there are overlapping matches.  How many matches are there if we search for "aba" in "ababababa"?